### PR TITLE
MAINT deprecate `self.N` and `self.J_pad` in 1D

### LINKED ISSUE #857: redefine 1D `self.N`, deprecate`self.J_pad`?
in 1D, `self.N` is the unpadded length of the input signal. It is defined in `ScatteringBase1D` as some derivative of `self.shape` and isn't being used anywhere else. 

Same for 2D where `self.shape` unpacks as `self.M`, `self.N`. 

I see no reason to keep these values as attributes in the object when:
* we have `self.shape` for that
* these values are only being used transiently, in `self.build()`
* our filterbank construction relies on `N_padded`, not `N`
* these values are trivially available in the forward via `self.N = len(x)`

In the spirit of preparing for scattering on variable-length inputs (`shape=None`, a long-requested feature !), i ask the permission to remove these attributes from `self`.

I am open to making a deprecation cycle where we convert these attributes to read-only properties and print a `DeprecationWarning` in v0.3 if anyone tries to access these numbers (NB: none of our examples access `self.N` and friends)

@janden @eickenberg

### PR DESCRIPTION
**Pull request by lostanlen**
Fixes #857 

An important deprecation to set before we release v0.3 is `J_pad`. Indeed, there won't be a `J_pad_fr` in JTFS, so it'd best to plan for symmetry in design.

```python
_N_padded, _N_padded_fr
J, J_fr
Q, Q_fr
T, F
log2_T, log2_F
```

I have moved the shape check to its own static method. We'll be able to call this method a second time inside `self._check_input(x)` (recently merged!)

### CODE REVIEW

File: tests/scattering1d/test_utils_scattering1d.py [hunk 1]
```diff
@@ -10,15 +10,15 @@ def test_compute_padding():
...
     assert pad_left == 8 and pad_right == 8
 
     with pytest.raises(ValueError) as ve:
-        _, _ = compute_padding(3, 16)
+        _, _ = compute_padding(8, 16)
     assert "should be larger" in ve.value.args[0]
 
     with pytest.raises(ValueError) as ve:
-        _, _ = compute_padding(6, 16)
+        _, _ = compute_padding(64, 16)
```
Comment (MuawizChaudhary) [inline tests/scattering1d/test_utils_scattering1d.py:21]: Why increase?

**Review by MuawizChaudhary — APPROVED** [review]
I approve.

File: tests/scattering1d/test_utils_scattering1d.py (hunk 1 above)
Comment (lostanlen) [reply tests/scattering1d/test_utils_scattering1d.py:21]: these are the same tests, but with the first argument being the exponent of 2.
before: J_pad=5, N_input=16
after: N=2**J_pad=32, N_input=16

etc.

**Comment by lostanlen** [comment]
next up, fix the newly introduced `DeprecationWarning` in `scattering_filter_factory` !

**Comment by lostanlen** [comment]
i have reduced the PR size. this runtime shape check i experimented with should happen in a separate PR, ideally a backend agnostic one

**Comment by lostanlen** [comment]
seems like i've broken everything with this commit. i can't even run the tests anymore 🤕

**Comment by janden** [comment]
Looks like the OOM killer strikes again: https://stackoverflow.com/questions/68844666/github-action-is-being-killed…

**Comment by lostanlen** [comment]
Facepalm. I guess building a filterbank of support `2**(2**J)` wasn't too responsible of me

**Comment by lostanlen** [comment]
Alright, ready now. I'm waiting for the codecov report and then we can consider merging.
Some open PRs will need a rebase

**Comment by lostanlen** [comment]
i'm making a public service announcement (extended to @cyrusvahidi @changhongw) because this PR has an important change of semantics. In your code:

* replace `self.N` by `self.shape[0]` (input length)
* replace `2**self.J_pad` by `len(self.phi[0])` (filter length)

**Comment by lostanlen** [comment]
rebased. can i get approval from either @changhongw or @cyrusvahidi ?

**Comment by cyrusvahidi** [comment]
> rebased. can i get approval from either @changhongw or @cyrusvahidi ?

LGTM

File: kymatio/scattering1d/frontend/base_frontend.py [hunk 2]
```diff
@@ -40,46 +40,48 @@ def build(self):
...
 
-        # check the shape
+        # check input length
         if isinstance(self.shape, numbers.Integral):
-            self.N = self.shape
+            self.shape = (self.shape,)
         elif isinstance(self.shape, tuple):
-            self.N = self.shape[0]
-            if len(self.shape) > 1:
+            if (len(self.shape)>1):
```
Comment (cyrusvahidi) [inline kymatio/scattering1d/frontend/base_frontend.py:47]: pedantic but `len(self.shape) > 1:` please

Comment (lostanlen) [reply kymatio/scattering1d/frontend/base_frontend.py]: 9478260

**Review by cyrusvahidi — APPROVED** [review]

File: kymatio/scattering1d/filter_bank.py [hunk 3]
```diff
@@ -434,16 +435,14 @@ def scattering_filter_factory(J_support, J_scattering, Q, T, r_psi=math.sqrt(0.5
...
-                j1 for j1 in j1s if j2 > j1]
+            possible_subsamplings_after_order1 = [j1 for j1 in j1s if j2 > j1]
             if len(possible_subsamplings_after_order1) > 0:
                 max_sub_psi2 = max(possible_subsamplings_after_order1)
             else:
                 max_sub_psi2 = 0
         else:
             max_sub_psi2 = max_subsampling
         # We first compute the filter without subsampling
-        N = 2**J_support
```
Comment (MuawizChaudhary) [inline kymatio/scattering1d/filter_bank.py:446]: We do the `2**J_support` outside of this call now?

File: kymatio/scattering1d/frontend/numpy_frontend.py [hunk 4]
```diff
@@ -3,6 +3,7 @@
 from ...frontend.numpy_frontend import ScatteringNumPy
 from ..core.scattering1d import scattering1d
 from .base_frontend import ScatteringBase1D
+import numpy as np
```
Comment (MuawizChaudhary) [inline kymatio/scattering1d/frontend/numpy_frontend.py:6]: Why?

**Review by MuawizChaudhary — CHANGES_REQUESTED** [review]

File: kymatio/scattering1d/filter_bank.py (hunk 3 above)
Comment (lostanlen) [reply kymatio/scattering1d/filter_bank.py:446]: we now pass `N` as first argument. no need to compute it inside the function anymore

File: kymatio/scattering1d/frontend/numpy_frontend.py (hunk 4 above)
Comment (lostanlen) [reply kymatio/scattering1d/frontend/numpy_frontend.py:6]: oops! a remnant of a previous commit

**Review by MuawizChaudhary — APPROVED** [review]
Looking to @janden for a final approval.

**Review by cyrusvahidi — APPROVED** [review]

**Comment by lostanlen** [comment]
@MuawizChaudhary i think @janden already approved the overall idea, and we talked about that during the meeting yesterday. I think i can speak for him that either you or @cyrusvahidi are free to merge it

**Comment by cyrusvahidi** [comment]
> @MuawizChaudhary i think @janden already approved the overall idea, and we talked about that during the meeting yesterday. I think i can speak for him that either you or @cyrusvahidi are free to merge it

can you resolve the conflicts?

**Comment by lostanlen** [comment]
done

**Comment by lostanlen** [comment]
Rebased from #854, which was relying on the old semantics (`2**J_scattering`)

@MuawizChaudhary please, can we make this PR next in line for merging?

**Review by cyrusvahidi — APPROVED** [review]

**Comment by MuawizChaudhary** [comment]
The pull request fulfills the specifications laid out by @lostanlen and @janden in issue #857 . Therefore I am merging.

If there are any complaints about the design proposed here, please make an issue or discuss it during our next weekly meeting.

**Review by janden — COMMENTED** [review]
Looks good but we're now missing docstring for `N`, `J` in `scattering_filter_factory`.

**Comment by lostanlen** [comment]
oops! yes good catch, needs a new doc now.

thanks for merging 🙏🏻

**Pull request merged**